src/st/cat/e_hellfire_beast.c

```diff
@@ -0,0 +1,1189 @@
+        primIndex = g_api.AllocPrimitives(PRIM_GT4, 0x12);
+        if (primIndex != -1) {
+            self->flags |= FLAG_HAS_PRIMS;
+            self->primIndex = primIndex;
+            prim = &g_PrimBuf[primIndex];
+            self->ext.hellfireBeast.primThree = prim;
+
+            drEnv = g_api.func_800EDB08((POLY_GT4*)prim);
+            if (drEnv == NULL) {
+                DestroyEntity(self);
+                break;
+            }
+
+            prim->type = PRIM_ENV;
+            drawEnv = g_CurrentBuffer->draw;
+            drawEnv.isbg = 1;
+            drawEnv.clip = rect;
+            drawEnv.ofs[0] = 0;
+            drawEnv.ofs[1] = 0x100;
+            SetDrawEnv(drEnv, &drawEnv);
+            prim->priority = 0xFF;
+            prim->drawMode = DRAW_HIDE;
+            prim = prim->next;
+
+            drEnv = g_api.func_800EDB08((POLY_GT4*)prim);
+            if (drEnv == NULL) {
+                DestroyEntity(self);
+                break;
+            }
+
+            prim->type = PRIM_ENV;
+            prim->priority = 0x101;
+            prim->drawMode = DRAW_UNK_800 | DRAW_HIDE;
+            prim = prim->next;
+
+            self->ext.hellfireBeastFlamePillar.prim = prim;
+            while (prim != NULL) {
+                prim->tpage = 0x110;
+                prim->u0 = prim->u1 = 0;
+                prim->u2 = prim->u3 = 0x12;
+                prim->v0 = prim->v2 = 0;
+                prim->v1 = prim->v3 = 0x3F;
+#ifdef VERSION_PSP
+                PGREY(prim, 0) = 0xC0;
+#else
+                PGREY(prim, 0) = 0x60;
+#endif
+                LOW(prim->r1) = LOW(prim->r0);
+                LOW(prim->r2) = LOW(prim->r0);
+                LOW(prim->r3) = LOW(prim->r0);
+                prim->p3 = 0;
+                prim->priority = self->zPriority;
+                prim->drawMode = DRAW_HIDE;
+                prim = prim->next;
+            }
+        } else {
+            DestroyEntity(self);
+            break;
+        }
+
+        prim = self->ext.hellfireBeastFlamePillar.prim;
+        prim = FindFirstUnkPrim(prim);
+        if (prim != NULL) {
+            prim->p3 = 1;
+            prim->p1 = 0;
+            prim->y0 = prim->y1 = self->posY.i.hi;
+            prim->y2 = prim->y3 = self->posY.i.hi;
+            prim->drawMode = DRAW_TPAGE2 | DRAW_TPAGE | DRAW_COLORS |
+                             DRAW_UNK02 | DRAW_TRANSP;
+        }
+        self->ext.hellfireBeastFlamePillar.unk80 = 0x18;
+        self->ext.hellfireBeastFlamePillar.castTimer = -0x10;
+        PlaySfxPositional(SFX_FIREBALL_SHOT_A);
+        self->step++;
+
+        break;
+    case 3:
+        prim = self->ext.hellfireBeast.primThree;
+        prim->drawMode = DRAW_DEFAULT;
+
+        prim = prim->next;
+        prim->drawMode = DRAW_UNK_800;
+
+        // NB. Strange way to write a single loop iteration...
+        // Maybe the number of iterations changed during development?
+        for (i = 0; i <= 0; i++) {
```

Comment:
This is a bit weird but I tried rewriting it in other ways for a single iteration like `do { } while(0)` and it changed the asm. This works at least